components/net/http_loader.rs

```diff
@@ -1020,42 +1036,124 @@ fn http_network_or_cache_fetch(
+                    // Ensure the done chan is not set if we're not using the cached response,
+                    // as the cache might have set it to Some if it constructed a pending response.
+                    *done_chan = None;
+
+                    // Update the cache state, incrementing the pending store count,
+                    // or starting the count.
+                    if let HttpCacheEntryState::PendingStore(i) = *state {
+                        let new = i + 1;
+                        *state = HttpCacheEntryState::PendingStore(new);
+                    } else {
+                        *state = HttpCacheEntryState::PendingStore(1);
+                    }
                 }
-                if let Some(entity_tag) = response_headers.get(header::ETAG) {
-                    http_request
-                        .headers
-                        .insert(header::IF_NONE_MATCH, entity_tag.clone());
+                if needs_revalidation {
+                    revalidating_flag = true;
+                    // Substep 5
+                    if let Some(http_date) = response_headers.typed_get::<LastModified>() {
+                        let http_date: SystemTime = http_date.into();
+                        http_request
+                            .headers
+                            .typed_insert(IfModifiedSince::from(http_date));
+                    }
+                    if let Some(entity_tag) = response_headers.get(header::ETAG) {
+                        http_request
+                            .headers
+                            .insert(header::IF_NONE_MATCH, entity_tag.clone());
+                    }
+                } else {
+                    // Substep 6
+                    response = cached_response;
                 }
+            }
+        }
+        // Notify the next thread waiting in line, if there is any.
+        if *state == HttpCacheEntryState::ReadyToConstruct {
+            cvar.notify_one();
+        }
+        // End of critical section on http-cache state.
+    }
+
+    // Decrement the number of pending stores,
+    // and set the state to ready to construct,
+    // if no stores are pending.
+    fn update_http_cache_state(context: &FetchContext, http_request: &Request) {
+        let (lock, cvar) = {
+            let entry_key = CacheKey::new(&http_request);
+            let mut state_map = context.state.http_cache_state.lock().unwrap();
+            &*state_map
+                .get_mut(&entry_key)
+                .expect("Entry in http-cache state to have been previously inserted")
+                .clone()
+        };
+        let mut state = lock.lock().unwrap();
+        if let HttpCacheEntryState::PendingStore(i) = *state {
+            let new = i - 1;
+            if new == 0 {
+                *state = HttpCacheEntryState::ReadyToConstruct;
+                // Notify the next thread waiting in line, if there is any.
+                cvar.notify_one();
```

Comment:
I slipped-in on change, which is to wake-up only one thread here, and then add a call to notify the next in line at https://github.com/servo/servo/pull/24318/files#diff-aa469beb5619907dbccd88364264b9b8R1131 @asajeffrey